README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# emstreeR

<!-- # emstreeR <img src="man/figures/logo.png" align="right" /> -->
<!-- [![Downloads](http://cranlogs.r-pkg.org/badges/emstreeR?color=brightgreen)](http://www.r-pkg.org/pkg/emstreeR) -->
<!-- one space after links to display badges side by side -->
<!-- badges: start -->

<!-- [![Travis-CI Build Status](https://travis-ci.org/allanvc/emstreeR.svg?branch=master)](https://travis-ci.org/allanvc/emstreeR)  -->
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/emstreeR)](https://cran.r-project.org/package=emstreeR)
[![Downloads from the RStudio CRAN
mirror](https://cranlogs.r-pkg.org/badges/grand-total/emstreeR)](https://cran.r-project.org/package=emstreeR)
[![License](https://img.shields.io/badge/License-BSD3--Clause-blue?style=flat-square)](https://opensource.org/license/bsd-3-clause/)
[![R-CMD-check](https://github.com/allanvc/emstreeR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/allanvc/emstreeR/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

## Overview

`emstreeR` enables **R** users to fast and easily compute an Euclidean
Minimum Spanning Tree (EMST) from data. This package relies on the R API
for {mlpack} - the C++ Machine Learning Library (Curtin et. al., 2013).
{emstreeR} uses the Dual-Tree Boruvka (March, Ram, Gray, 2010,
<https://doi.org/10.1145/1835804.1835882>), which is theoretically and
empirically the fastest algorithm for computing an EMST. This package
also provides functions and an S3 method for readily plotting Minimum
Spanning Trees (MST) using either the style of the {base},
{scatterplot3d}, or {ggplot2} libraries; and functions to export the MST
output to shapefiles.

- `computeMST()` computes an Euclidean Minimum Spanning Tree for the
  input data.
- `plot.MST()` an S3 method for the generic function `plot()` that
  produces 2D MST plots.
- `plotMST3D()` plots a 3D MST using the {scatterplot3d} style.
- `stat_MST()` a {ggplot2} Stat extension for plotting a 2D MST.
- `export_vertices_to_shapefile()` writes a shapefile containing the
  `MST` vertices.
- `export_edges_to_shapefile()` writes a shapefile containing the `MST`
  edges.

## Installation

``` r
# CRAN version
install.packages("emstreeR")

# Dev version
if (!require('devtools')) install.packages('devtools')
devtools::install_github("allanvc/emstreeR")
```

## Basic Usage

``` r
## artificial data:
set.seed(1984)
n <- 7
c1 <- data.frame(x = rnorm(n, -0.2, sd = 0.2), y = rnorm(n, -2, sd = 0.2))
c2 <- data.frame(x = rnorm(n, -1.1, sd = 0.15), y = rnorm(n, -2, sd = 0.3)) 
d <- rbind(c1, c2)
d <- as.data.frame(d)

## MST:
library(emstreeR)
out <- ComputeMST(d)
out
```

    ##               x         y from to   distance
    ## 1  -0.118159357 -2.166545   11 13 0.03281747
    ## 2  -0.264604994 -2.105242    8 12 0.05703382
    ## 3  -0.072829535 -1.716803    3  7 0.08060398
    ## 4  -0.569225757 -1.943598    5  6 0.11944501
    ## 5  -0.009270527 -1.942413    6  7 0.13450475
    ## 6   0.037697969 -1.832590    8 10 0.14293342
    ## 7  -0.091509110 -1.795213    1  2 0.15875908
    ## 8  -1.097338236 -1.871078   10 14 0.16993335
    ## 9  -0.841400898 -2.194585    1  5 0.24918237
    ## 10 -1.081888729 -1.728982    8 13 0.27882008
    ## 11 -1.366334073 -2.003965    2  4 0.34485145
    ## 12 -1.081078171 -1.925745    9 12 0.36016689
    ## 13 -1.357063682 -1.972485    4  9 0.37023475
    ## 14 -0.913706515 -1.753315    1  1 0.00000000

## Plotting

### 2D Plots

``` r
## artifical data for 2D plots:
set.seed(1984)
n <- 15
c1 <- data.frame(x = rnorm(n, -0.2, sd = 0.2), y = rnorm(n, -2, sd = 0.2))
c2 <- data.frame(x = rnorm(n, -1.1, sd = 0.15), y = rnorm(n, -2, sd = 0.3)) 
d <- rbind(c1, c2)
d <- as.data.frame(d)
  
## MST:
library(emstreeR)
out <- ComputeMST(d, verbose = FALSE)
```

``` r
## simple 2D plot:
plot(out, col.pts = "red", col.segts = "blue")
```

<img src="man/README-figures/base-1.png" width="650" height="500">

``` r
## 2D plot with ggplot2:
library(ggplot2)
ggplot(data = out, aes(x = x, y = y, from = from, to = to))+ 
  geom_point()+ 
  stat_MST(colour="red")
```

<img src="man/README-figures/ggplot-1.png" width="600" height="400">

``` r
## 2D curved edges plot with ggplot2:
library(ggplot2)
ggplot(data = out, aes(x = x, y = y, from = from, to = to))+ 
  geom_point()+ 
  stat_MST(geom="curve")
```

<img src="man/README-figures/ggplot_curved-1.png" width="600" height="400">

### 3D Plot

``` r
## artificial data for 3D plots:
n = 99
set.seed(1984)
d1 <- matrix(rnorm(n, mean = -2, sd = .5), n/3, 3) # 3d
d2 <- matrix(rnorm(n, mean = 0, sd = .3), n/3, 3)
d3 <- matrix(rnorm(n, mean = 3, sd = .4), n/3, 3)
d <- rbind(d1,d2,d3) # showing a matrix input
  
## MST:
library(emstreeR)
out <- ComputeMST(d, verbose = FALSE)
```

``` r
## simple 3D plot:
plotMST3D(out, xlab = "xaxis", col.pts = "orange", col.segts = "red", main = "a simple MST 3D plot")
```

<img src="man/README-figures/scatterplot3d-1.png" width="600" height="400">

### Exporting the Output to GIS Shapefiles

``` r
## mock data
country_coords_txt <- "
1     3.00000  28.00000       Algeria
2    54.00000  24.00000           UAE
3   139.75309  35.68536         Japan
4    45.00000  25.00000 'Saudi Arabia'
5     9.00000  34.00000       Tunisia
6     5.75000  52.50000   Netherlands
7   103.80000   1.36667     Singapore
8   124.10000  -8.36667         Korea
9    -2.69531  54.75844            UK
10    34.91155  39.05901        Turkey
11  -113.64258  60.10867        Canada
12    77.00000  20.00000         India
13    25.00000  46.00000       Romania
14   135.00000 -25.00000     Australia
15    10.00000  62.00000        Norway"
 
d <- read.delim(text = country_coords_txt, header = FALSE,
                 quote = "'", sep = "",
                 col.names = c('id', 'lon', 'lat', 'name'))
                 
## MST
library(emstreeR)
output <- ComputeMST(d[,2:3])
#plot(output)

export_vertices_to_shapefile(output, file="vertices.shp")
export_edges_to_shapefile(output, file="edges.shp")
```

Below is an example of how to open the shapefiles using the QGIS
software in Ubuntu.

Open file `vertices.shp` or `edges.shp`.

<img src="man/README-figures/qgis1.png" width="750" height="400">

<img src="man/README-figures/qgis2.png" width="750" height="400">

Then go to `Menu > Layer > Add Layer > Add Vector Layer`.

<img src="man/README-figures/qgis3.png" width="750" height="400">

Select `Source Type` as `File` if it is not selected yet. Then click on
the three dots button under `Source` to select the other shapefile,
depending on which one you used to open QGIS. In the example below, we
select `vertices.shp` as we chose `edges.shp` first.

<img src="man/README-figures/qgis4.png" width="750" height="400">

<img src="man/README-figures/qgis5.png" width="750" height="400">

Hit `Add`, then `Close` and **voilà**.

<img src="man/README-figures/qgis6.png" width="750" height="400">

It is then very straightforward to add other layers such map shapefiles
or add the generated EMST to existing layers.

## License

This package is licensed under the terms of the BSD 3-clause License.

## References

March, W. B., and Ram, P., and Gray, A. G. (2010). *Fast euclidian
minimum spanning tree: algorithm analysis, and applications*. 16th ACM
SIGKDD International Conference on Knowledge Discovery and Data mining,
July 25-28 2010. Washington, DC, USA.

Curtin, R. R. et al. (2013). Mlpack: A scalable C++ machine learning
library. *Journal of Machine Learning Research*, v. 14, 2013.


NEWS.md
# emstreeR (2023-11-13 - CRAN update)

* carries all the updates after 3.0.0 to CRAN
* removed emstreeR-package.Rd to avoid CRAN notes
* incorporates new vignette showing how to export shapefiles

# emstreeR 3.1.0 (2023-11-07 - Github patch)

## doc 

* updated DESCRIPTION and README


## new features

* included support to export to export output to shapefiles (new functions: `export_vertices_to_shapefile()` and `export_edges_to_shapefile()`)

## bug fix

* `plot.MST` was not being exported. Included #'@export in the documentation file

## others

* removed dependency BBMisc

# emstreeR 3.0.0 (2020-03-17 - Github patch/CRAN update)

* Migrated from RcppMLPACK to mlpack [thanks to @nacnudus] [#3]

## doc

* updated DESCRIPTION and README


# emstreeR 2.2.2 (2020-11-29 - Github patch/CRAN update)

## functions

* data_check() and mlpack_mst's doc are now @noRd

## doc

* fixed typos in DESCRIPTION and README

* incorporated the pull request 1

## bug fix

* fixed the documentation (.Rd) file of `computeMST()`. A wrong link was causing the malfunction in the dev version.

---

# emstreeR 2.2.1 (2019-08-20 - Github patch)

## functions

* data_check() and mlpack_mst are now internal
* changed .Rbuildignore to ignore README.Rmd, NEWS.md, cran-comments.md and the
/inst folder

---

# emstreeR 2.2.0 (2019-05-07 - 4th CRAN patch)

## New data, functions, and features

* added 'scale' parameter in the ComputeMST() function. It gives the option to 
scale or not your data before computing the MST (Thank's to Moses Obiri who 
pointed this out).
* updated documentation in ComputeMST.R file
* updated documentation in emstreeR-package.Rd file
* deleted honeymoon example (extras) from README file, since dsk service for maps are not
available anymore.

---

# emstreeR 2.1.2 (2019-03-13 - third CRAN patch)

## New data, functions, and features

* updated DESCRIPTION file.
* fixed citations in DESCRIPTION, README and emstreeR-package.Rd
* updated documentation in the ComputeMST.R file 

---

# emstreeR 2.1.1 (2019-01-19 - second CRAN patch)

## Internal

* Fixed Imports in DESCRIPTION file, removing the mention to RcppMLPACK in order
to eliminate NOTES when building on Fedora and Solaris

## New data, functions, and features

* updated DESCRIPTION file.
* updated 'emstreeR-package' file.
* updated 'README.md' file.

---

# emstreeR 2.1.0 (2018-12-06 - first CRAN patch)

## Internal

* Fixed compilation in OS X and Solaris:
  - copied 'Makervars' and 'Makevars.win' from RcppMLPACK with minor modifications
  - added //Rcpp[[depends(RcppMLPACK)]] to mlpack_mst.cpp file

## New data, functions, and features

* fixed typos in the DESCRIPTION file.

---

# emstreeR 2.0.0 (2018-11-28 - pre CRAN release)

## Internal

* Fix [-Wlang-lang] windows compilation WARNINGs when running devtools::win_check() 
  by adding `CXX_STD = CXX_11` line to the Makevars.win file. Reference: "Writing
  R Extensions" manual, section 1.2.4 - "Using C++11 code".

## New data, functions, and features

* Substituted the wrapper function `plotMST()` for the S3 method `plot.MST()`

* Added 'dontrun' example 'honeymoon cruise' to the package README file 

* Removed broken Japan 'dontrun' example from the package help page 
  'emstreeR-package.Rd'because of the token problem with GoogleMaps API.

* fixed errors in the DESCRIPTION file as suggested by CRAN


R/stat-MST.R
#' @title Euclidean Minimum Spanning Tree Stat Function
#'
#' @description A Stat extension for 'ggplot2' to plot a 2D MST by making a 
#'     scatter plot with segments.
#' @description \code{stat_MST} uses the information returned by 
#'    \code{\link[emstreeR]{ComputeMST}} for producing a 2D Minimum Spanning 
#'    Tree plot with 'ggplot2' and should be combined with \code{geom_point()}.
#'    
#' @param mapping The aesthetic mapping, usually constructed with
#'   \code{\link[ggplot2]{aes}} or \code{\link[ggplot2]{aes_}}. 
#'    The required aesthetics are \code{x}, \code{y}, \code{from}, and \code{to}. 
#'    Those are columns of the \code{mst} object returned by 
#'    \code{\link[emstreeR]{ComputeMST}}.
#' @param data a \code{mst} class object returned by the 
#'    \code{\link[emstreeR]{ComputeMST}} function.
#' @param geom The geometric object to display the data. The default value is
#'    \code{"segment"} in order to produce the edges between the vertices.
#' @param position The position adjustment to use for overlapping points on this
#'    layer
#' @param show.legend logical. Should this layer be included in the legends?
#'    \code{NA}, the default, includes if any aesthetics are mapped. \code{FALSE}
#'    never includes, and \code{TRUE} always includes.
#' @param inherit.aes If \code{FALSE}, overrides the default aesthetics, rather
#'    than combining with them. This is most useful for helper functions that
#'    define both data and aesthetics and shouldn't inherit behaviour from the
#'    default plot specification, e.g. \code{\link[ggplot2]{borders}}.
#' @param linetype an integer or name: \code{0 = "blank"}, \code{1 = "solid"},
#'    \code{2 = "dashed"}, \code{3 = "dotted"}, \code{4 = "dotdash"}, 
#'    \code{5 = "longdash"}, \code{6 = "twodash"}. The default for \code{'MST'} 
#'    objects is "dotted".
#' @param ... other arguments passed on to \code{\link[ggplot2]{layer}}. This
#'    can include aesthetics whose values you want to set, not map. See
#'    \code{\link[ggplot2]{layer}} for more details.
#' @param na.rm	a logical value indicating whether \code{NA} values should be 
#' stripped before the computation proceeds.
#'
#' @section Computed variables: \describe{ 
#'    \item{x}{x coordinates of the MST start points}
#'    \item{y}{y coordinates of the MST start points}
#'    \item{xend}{x coordinates of the MST end points}
#'    \item{yend}{y coordinates of the MST end points}
#'    }
#'
#' @examples
#' 
#' ## 2D artificial data:
#' set.seed(1984)
#' n <- 15
#' c1 <- data.frame(x = rnorm(n, -0.2, sd = 0.2), y = rnorm(n, -2, sd = 0.2))
#' c2 <- data.frame(x = rnorm(n, -1.1, sd = 0.15), y = rnorm(n, -2, sd = 0.3)) 
#' d <- rbind(c1, c2)
#' d <- as.data.frame(d)
#' 
#' ## MST:
#' out <- ComputeMST(d)
#' 
#' #1) simple plot
#' library(ggplot2)
#' ggplot(data = out, 
#'     aes(x = x, y = y, 
#'     from = from, to = to))+ 
#'     geom_point()+
#'     stat_MST(colour = "red", linetype = 2)
#'     
#' #2) curved edges
#' library(ggplot2)
#' ggplot(data = out, 
#'     aes(x = x, y = y, 
#'     from = from, to = to))+ 
#'     geom_point()+
#'     stat_MST(geom = "curve", colour = "red", linetype = 2)
#'
#' \dontrun{
#' ## plotting MST on maps:
#' library(ggmap)
#' 
#' #3) honeymoon cruise example
#' # define ports
#' df.port_locations <- data.frame(location = c("Civitavecchia, Italy",
#'                                              "Genova, Italy",
#'                                              "Marseille, France",
#'                                              "Barcelona, Spain",
#'                                              "Tunis, Tunisia",
#'                                              "Palermo, Italy"), 
#'                                 stringsAsFactors = FALSE)
#' 
#' # get latitude and longitude
#' geo.port_locations <- geocode(df.port_locations$location, source = "dsk")
#' 
#' # combine data
#' df.port_locations <- cbind(df.port_locations, geo.port_locations)
#' 
#' # MST
#' out <- ComputeMST(df.port_locations[,2:3])
#' plot(out) #just to check
#' 
#' # Plot
#' #' map <- c(left = -8, bottom = 32, right = 20, top = 47)
#' 
#' get_stamenmap(map, zoom = 5) %>% ggmap()+
#'   stat_MST(data = out, 
#'            aes(x = lon, y = lat, from = from, to = to), 
#'            colour = "red", linetype = 2)+
#'   geom_point(data = out, aes(x = lon, y = lat), size = 3)
#' 
#' 
#' #4) World Map travels:
#' library(ggplot2)
#' library(ggmaps)
#' 
#' country_coords_txt <- "
#'    1     3.00000  28.00000       Algeria
#'    2    54.00000  24.00000           UAE
#'    3   139.75309  35.68536         Japan
#'    4    45.00000  25.00000 'Saudi Arabia'
#'    5     9.00000  34.00000       Tunisia
#'    6     5.75000  52.50000   Netherlands
#'    7   103.80000   1.36667     Singapore
#'    8   124.10000  -8.36667         Korea
#'    9    -2.69531  54.75844            UK
#'    10    34.91155  39.05901        Turkey
#'    11  -113.64258  60.10867        Canada
#'    12    77.00000  20.00000         India
#'    13    25.00000  46.00000       Romania
#'    14   135.00000 -25.00000     Australia
#'    15    10.00000  62.00000        Norway"
#'    
#'    
#'  d <- read.delim(text = country_coords_txt, header = FALSE, 
#'    quote = "'", sep = "", col.names = c('id', 'lon', 'lat', 'name'))
#'    
#'  out <- ComputeMST(d[,2:3])
#'  
#'  country_shapes <- geom_polygon(aes(x = long, y = lat, group = group),
#'    data = map_data('world'), fill = "#CECECE", color = "#515151", 
#'    size = 0.15)
#'    
#'  ggplot()+ country_shapes+
#'    stat_MST(geomdata = out, aes(x = lon, y = lat, from = from, to = to), 
#'      colour = "red", linetype = 2)+
#'    geom_point(data = out, aes(x = lon, y = lat), size=2)
#'}
#'
#' @export
#' 
stat_MST <- function(mapping = NULL, data = NULL, geom = "segment", 
                     position = "identity", na.rm = FALSE, 
                     linetype="dotted", show.legend = NA, 
                     inherit.aes = TRUE, ...) {
  ggplot2::layer(
    stat = StatMST, 
    data = data, 
    mapping = mapping, 
    geom = geom, 
    position = position, 
    show.legend = show.legend, 
    inherit.aes = inherit.aes,
    params = list(na.rm = na.rm, linetype = linetype, ...)
  )
}

#' @title 'ggproto' objects
#'
#' @description Internal use only definitions of 'ggproto' objects needed for
#'   geoms and stats.
#'
#' @name emstreeR-ggproto
#' @format NULL
#' @usage NULL
#' @keywords internal
NULL
#> NULL

#' @rdname emstreeR-ggproto
#' @format NULL
#' @usage NULL
#' @export
StatMST <- ggplot2::ggproto("StatMST", ggplot2::Stat,
                       compute_group = function(data, scales, ...) {
                         ## Compute the line segment endpoints
                         
                         # change the last row's from/to columns
                         # repeat the row above
                         # this is done in order to work with geom="curve"
                         # avoiding the Error in calcCurveGrob(x, x$debug) : 
                         #.. end points must not be identical
                         data[nrow(data), c("from", "to")] <- 
                           c(
                             data[1, "from"], data[1, "to"]
                             )
                         
                         x = data[data$from, 1]
                         y = data[data$from, 2]
                         xend = data[data$to, 1]
                         yend = data[data$to, 2]
                         
                         data.frame(x=x, y=y, xend=xend, yend=yend)
                         
                       },
                       required_aes = c("x", "y", "from", "to")
)
